python/glow/wgr/wgr_functions.py

```diff
@@ -199,3 +204,46 @@ def transpose_one(contig):
         values = list(transposed_df.to_numpy())
     transposed_df['values_array'] = values
     return spark.createDataFrame(transposed_df[['values_array']].reset_index(), column_names)
+
+
+def estimate_loco_offsets(block_df: DataFrame,
+                          label_df: pd.DataFrame,
+                          sample_blocks: Dict[str, List[str]],
+                          cov_df: pd.DataFrame = pd.DataFrame({}),
+                          add_intercept: bool = True,
+                          alphas: NDArray[(Any, ), Float] = np.array([]),
+                          label_type='detect',
+                          chromosomes: List[str] = []) -> pd.DataFrame:
+    """
+    The one-stop function to generate WGR predictors to be used as offsets in gwas functions. Given the
+    input the function performs the ridge reduction followed by appropriate choice of ridge regression or
+    logistic ridge regression in a loco manner.
+
+    Args:
+        block_df : Spark DataFrame representing the beginning block matrix X
+        label_df : Pandas DataFrame containing the target labels used in fitting the ridge models
+        sample_blocks : Dict containing a mapping of sample_block ID to a list of corresponding sample IDs
+        cov_df : Pandas DataFrame containing covariates to be included in every model in the stacking
+            ensemble (optional).
+        add_intercept: If True, an intercept column (all ones) will be added to the covariates
+            (as the first column)
+        alphas : array_like of alpha values used in the ridge reduction (optional).
+        label_type: String to determine type treatment of labels. It can be 'detect' (default), 'binary',
+            or 'quantitative'.
+        chromosomes : List of chromosomes for which to generate offsets (optional). If not provided, the
+                chromosomes will be inferred from the block matrix.
+
+    Returns:
+        Pandas DataFrame containing offset values per chromosome. The rows are indexed by sample ID and
+        chromosome; the columns are indexed by label. The column types are float64. The DataFrame is sorted using
+        chromosome as the primary sort key, and sample ID as the secondary sort key.
+    """
+    ridge_reduced = RidgeReduction(block_df, label_df, sample_blocks, cov_df, add_intercept, alphas,
+                                   label_type)
+    ridge_reduced.fit_transform()
```

Comment:
We probably want to cache this so it's not computed twice. We cache in our examples.